rustls/src/server/verifier_builder.rs

```diff
@@ -0,0 +1,318 @@
+}
+
+impl ClientCertVerifierBuilder {
+    pub(crate) fn new() -> Self {
+        Self {
+            roots: None,
+            crls: Vec::new(),
+            anon_policy: AnonymousClientsPolicy::Forbidden,
+        }
+    }
+
+    /// Verify that if a client offers a client certificate, that it is issued by a trust anchor
+    /// present in the provided [RootCertStore]. Calling `with_roots` multiple times will merge
+    /// the trust anchors.
+    ///
+    /// All clients must provide a client certificate unless you have allowed unauthenticated
+    /// clients by calling [allow_unauthenticated()][ClientCertVerifierBuilder::allow_unauthenticated]
+    /// on the builder.
+    pub fn with_roots(self, roots: RootCertStore) -> Self {
+        Self {
+            roots: match self.roots {
+                None => Some(roots),
+                Some(mut existing) => {
+                    existing.add_server_trust_anchors(roots.roots.into_iter());
+                    Some(existing)
+                }
+            },
+            crls: self.crls,
+            anon_policy: self.anon_policy,
+        }
+    }
+
+    /// Verify the revocation state of presented client certificates against the provided
+    /// certificate revocation lists (CRLs). Calling `with_crls` multiple time will merge the
+    /// CRLs.
+    ///
+    /// Returns a [CertRevocationListError] if any of the provided CRLs are invalid.
+    pub fn with_crls(
+        mut self,
+        crls: impl IntoIterator<Item = UnparsedCertRevocationList>,
+    ) -> Result<Self, CertRevocationListError> {
+        self.crls.append(
+            &mut crls
+                .into_iter()
+                .map(|der_crl| der_crl.parse())
+                .collect::<Result<Vec<_>, CertRevocationListError>>()?,
+        );
+        Ok(Self {
+            crls: self.crls,
+            roots: self.roots,
+            anon_policy: self.anon_policy,
+        })
+    }
+
+    /// Allow unauthenticated clients to connect.
+    ///
+    /// If trust anchors have been provided using [with_roots()][ClientCertVerifierBuilder::with_roots]
+    /// then both clients that offer a client certificate issued by a trusted root, or clients
+    /// that offer no client certificate will be allowed to connect.
+    pub fn allow_unauthenticated(self) -> Self {
+        Self {
+            roots: self.roots,
+            crls: self.crls,
+            anon_policy: AnonymousClientsPolicy::Allowed,
+        }
+    }
+
+    /// Build a client certificate verifier.
+    ///
+    /// Once built, the provided `Arc<dyn ClientCertVerifier>` can be used with a Rustls
+    /// [crate::server::ServerConfig] to configure client certificate validation using
+    /// [`with_client_cert_verifier`][crate::ConfigBuilder<ClientConfig, WantsVerifier>::with_client_cert_verifier].
+    pub fn build(self) -> Result<Arc<dyn ClientCertVerifier>, ClientCertVerifierBuilderError> {
+        if !self.crls.is_empty() && self.roots.is_none() {
```

Comment:
`roots.is_none()` means `with_roots()` wasn't called, but I think we could usefully catch calling `with_roots` with an empty `RootCertStore`? maybe `self.roots` could stop being an `Option`, and this could become `self.roots.empty()`?

Reply:
Good idea, reworked.